### PR TITLE
Add missing hashbrown drain APIs for Map and Set

### PR DESCRIPTION
**Pull request by pedromfedricci**
Related to #4 - "The Drain iterators".

### CODE REVIEW

**Comment by pedromfedricci** [comment]
Hmm guess tarpaulin is compiling tests with rustc < 1.52.0?
Related to this change: [Eventually adopt rustdoc::all change](https://github.com/jonhoo/griddle/commit/013a92e2e73e4d8bdce7e4614062eb2738b2f8e6)

**Comment by jonhoo** [comment]
Hmm, I didn't think tarpaulin was using an old compiler..
@xd009642 do you know what might be causing [this error](https://dev.azure.com/jonhoo/jonhoo/_build/results?buildId=2211&view=logs&j=5d95feb4-002d-53af-5cf8-49e3eceb086d&t=cb16c9f5-fcda-53ec-f779-fee3de200d0e&l=50)?
```
   Compiling griddle v0.5.1 (/__w/1/s)
error: Broken pipe (os error 32)
warning: build failed, waiting for other jobs to finish...
thread 'main' panicked at 'already borrowed: BorrowMutError', src/tools/cargo/src/cargo/util/config/mod.rs:307:20
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
May 31 18:59:56.938 ERROR cargo_tarpaulin: Failed to compile tests! Error: griddle: an unknown tool name found in scoped lint: `rustdoc::all`
```

File: src/set.rs [hunk 1]
```diff
@@ -1818,6 +1979,45 @@ mod test_set {
...
+        let mut s = HashSet::<i32>::new();
+        drop(s.drain());
+        assert!(s.is_empty());
+    }
+
+    #[test]
+    fn test_drain() {
+        let mut s: HashSet<_> = (1..100).collect();
+
+        // try this a bunch of times to make sure we don't screw up internal state.
```
Comment (jonhoo) [inline src/set.rs:2003]: I'm actually not so concerned about repeated runs, but more about if the map (or set) is in various stages of refill. That is, I want to make sure that drain doesn't do anything weird when the "rest" map is still non-empty for example. Could you add a test that specifically checks both a map that _isn't_ split and one that is? I also want to make sure that even if a `Drain` is forgotten (with `mem::forget`), nothing bad happens when the map is then dropped.

**Comment by xd009642** [comment]
I'm not sure I'd have to look at the tool attributes and CI jobs, but I can say tarpaulin just uses the version of the compiler installed in the system as it calls out to the process. So either the rust version used in the tests is old or there's something else afoot 🤔 . I raised an issue in tarpaulin and I'll try to have a deeper look this week

**Comment by jonhoo** [comment]
So, this is specifically running with the Docker image (specifically `latest`), so there won't be anything in the environment beyond what that sets up.

**Comment by xd009642** [comment]
Okay that will be the latest stable compiler (at time the image was built) then

**Comment by jonhoo** [comment]
That's weird though, as the latest stable compiler _should_ support `rustdoc::all`, which seems to be what it's failing on :thinking:

File: src/set.rs (hunk 1 above)
Comment (pedromfedricci) [reply src/set.rs:2003]: ok will do!

**Comment by xd009642** [comment]
@jonhoo latest hadn't been updated since 1.51.0, a new `tarpaulin:latest` is now built and pushed with 1.52.0 👍

**Comment by jonhoo** [comment]
That seems to have done it, thanks!

@pedromfedricci Just the change from the comment left now and then I think we're :+1:

**Comment by pedromfedricci** [comment]
Hey! @jonhoo 
I've made a few adjustments, do you see anything else I'm missing?
ty

**Review by jonhoo — APPROVED** [review]
This is excellent, thank you!

**Pull request merged**